galasa-managers-parent/galasa-managers-core-parent/dev.galasa.textscan.manager/src/main/java/dev/galasa/textscan/spi/TextScannerImpl.java

```diff
@@ -16,88 +23,255 @@
+			}
+		}
+		if(count<1) {
+			throw new InvalidParameterException("Invalid occurancies number");
+		}
+		Matcher m = searchPattern.matcher(text);
+		int found =0;
+		int occurance = count;
+		boolean find = false;
+		while (occurance > 0) {
+			find = m.find();
+			if (!find&& found ==0) {
+				throw new MissingTextException("Search Pattern '"+searchPattern+"' was not found");
+			}
+			else if(!find) {
+				throw new IncorrectOccurancesException("Was looking for "+count+" instances of '"+searchPattern+"' but found "+ found);
+			}
+			found++;
+			occurance--;
+		}
+		return this;
+	}
+
+	@Override
+	public ITextScanner scan(String text, String searchString, String failString, int count)
+			throws FailTextFoundException, MissingTextException, IncorrectOccurancesException {
+
+		Pattern fp = Pattern.compile("\\Q" + failString + "\\E",Pattern.DOTALL | Pattern.MULTILINE);
+		Pattern p = Pattern.compile("\\Q" + searchString + "\\E",Pattern.DOTALL | Pattern.MULTILINE);
+
+		return scan(text,p,fp,count);
+
+	}
+
+	@Override
+	public ITextScanner scan(ITextScannable scannable, Pattern searchPattern, Pattern failPattern, int count)
+			throws FailTextFoundException, MissingTextException, IncorrectOccurancesException, ManagerException, IOException {
+		if(scannable.isScannableInputStream()) {
+			return scan(scannable.getScannableInputStream(),searchPattern,failPattern,count);	
+		}else if(scannable.isScannableString()) {
+			return scan(scannable.getScannableString(),searchPattern,failPattern,count);
+		}
+		/** It should never reach this line, it should either be an Input Steam or String **/
+		return null;
+	}
+
+	@Override
+	public ITextScanner scan(ITextScannable scannable, String searchString, String failString, int count)
+			throws FailTextFoundException, MissingTextException, IncorrectOccurancesException, ManagerException, IOException {
+		if(scannable.isScannableInputStream()) {
+			return scan(scannable.getScannableInputStream(),searchString,failString,count);	
+		}else if(scannable.isScannableString()) {
+			return scan(scannable.getScannableString(),searchString,failString,count);
+		}
+		/** It should never reach this line, it should either be an Input Steam or String **/
+		return null;
+	}
+
+	@Override
+	public ITextScanner scan(InputStream inputStream, Pattern searchPattern, Pattern failPattern, int count)
+			throws FailTextFoundException, MissingTextException, IncorrectOccurancesException, IOException {
+		BufferedReader  reader = new BufferedReader (new InputStreamReader(inputStream));
+		CircularFifoQueue<String> queue = new CircularFifoQueue<String>(10);
+		String line = "";
+		String toScan = "";
+		int found =0;
+		boolean find = false;
+		int lines = 0;
+		int occurance = 0;
+		if(count<1) {
+			throw new InvalidParameterException("Invalid occurancies number");
+		}
+		while((line =reader.readLine())!=null) {
+			occurance = count;
+			queue.add(line);
+			if(queue.isAtFullCapacity()) {
```

Comment:
simplify the code,  you don't have to wait for the queue to fill before you start checking